ndcube/ndcube.py

```diff
@@ -31,85 +32,199 @@
 __all__ = ['NDCubeABC', 'NDCubeBase', 'NDCube']
 
 
-class NDCubeABC(astropy.nddata.NDData, metaclass=abc.ABCMeta):
+class NDCubeABC(astropy.nddata.NDDataBase):
 
-    @abc.abstractproperty
-    def dimensions(self):
+    @property
+    @abc.abstractmethod
+    def extra_coords(self) -> ExtraCoordsABC:
         """
-        The array dimensions of the cube.
+        Coordinates not described by ``NDCubeABC.wcs`` which vary along one or more axes.
         """
 
+    @property
     @abc.abstractmethod
-    def crop(self, *points, wcs=None):
+    def global_coords(self) -> GlobalCoordsABC:
         """
-        Crop to the smallest cube in pixel space containing the world coordinate points.
+        Coordinate metadata which applies to the whole cube.
+        """
+
+    @property
+    @abc.abstractmethod
+    def combined_wcs(self) -> BaseHighLevelWCS:
+        """
+        The WCS transform for the NDCube, including the coordinates specified in ``.extra_coords``.
+
+        This transform should implement the high level wcsapi, and have
+        `pixel_n_dim` equal to the number of array dimensions in the
+        `.NDCube`. The number of world dimensions should be equal to the
+        number of world dimensions in ``self.wcs`` and in ``self.extra_coords`` combined.
+        """
+
+    @property
+    @abc.abstractmethod
+    def array_axis_physical_types(self) -> Iterable[Tuple[str, ...]]:
+        """
+        Returns the WCS physical types that vary along each array axis.
+
+        Returns an iterable of tuples where each tuple corresponds to an array axis and
+        holds strings denoting the WCS physical types associated with that array axis.
+        Since multiple physical types can be associated with one array axis, tuples can
+        be of different lengths. Likewise, as a single physical type can correspond to
+        multiple array axes, the same physical type string can appear in multiple tuples.
+
+        The physical types returned by this property are drawn from the
+        `~NDCube.combined_wcs` property so they include the coordinates contained in
+        `~.NDCube.extra_coords`.
+        """
+
+    @abc.abstractmethod
+    def axis_world_coords(self,
+                          *axes: Union[int, str],
```

Comment:
This is in the SEP: https://github.com/sunpy/sunpy-SEP/blob/master/SEP-0012.md

Reply:
Well I ask since I was told I was not allowed to add any typing to any code base in sunpy without a SEP. 

We had this conversation on sunpy-soar and the PR was closed for example.